src/ensembl/src/shared/tooltip/tooltip-helper.ts

```diff
@@ -0,0 +1,137 @@
+import {
+  TIP_WIDTH,
+  TIP_HEIGHT,
+  TIP_HORIZONTAL_OFFSET
+} from './tooltip-constants';
+
+import { Position } from './tooltip-types';
+
+const topRow = [Position.TOP_LEFT, Position.TOP_RIGHT];
+
+const bottomRow = [Position.BOTTOM_LEFT, Position.BOTTOM_RIGHT];
+
+const leftSide = [Position.LEFT_TOP, Position.LEFT_BOTTOM];
+
+const rightSide = [Position.RIGHT_TOP, Position.RIGHT_BOTTOM];
+
+type FindOptimalPositionParams = {
+  intersectionEntry: IntersectionObserverEntry;
+  anchorBoundingRect: ClientRect;
+  position: Position;
+};
+
+export const findOptimalPosition = (params: FindOptimalPositionParams) => {
+  const {
+    intersectionEntry: { isIntersecting }
+  } = params;
+  let { position } = params;
+
+  if (isIntersecting) {
+    return position;
+  }
+
+  return adjustPosition(params);
+};
+
+const adjustPosition = (params: FindOptimalPositionParams) => {
+  const possiblePositions = getPossiblePositions(params);
+  for (let position of possiblePositions) {
+    if (willStayWithinBounds({ ...params, position })) {
+      return position;
+    }
+  }
+
+  // if no better position was found, return original position
+  return params.position;
+};
+
+const getPossiblePositions = (params: FindOptimalPositionParams) => {
+  const { position } = params;
+  if (topRow.includes(position)) {
+    return [...topRow, ...bottomRow];
+  } else if (bottomRow.includes(position)) {
+    return [...bottomRow, ...topRow];
+  } else if (leftSide.includes(position)) {
+    return [...leftSide, ...rightSide];
+  } else if (rightSide.includes(position)) {
+    return [...rightSide, ...leftSide];
+  } else {
+    return [];
+  }
+};
+
+const willStayWithinBounds = (params: FindOptimalPositionParams) => {
+  const {
+    intersectionEntry: { boundingClientRect, rootBounds },
+    anchorBoundingRect,
+    position
+  } = params;
+
+  const {
+    left: anchorLeft,
+    right: anchorRight,
+    width: anchorWidth,
+    top: anchorTop,
+    bottom: anchorBottom,
+    height: anchorHeight
+  } = anchorBoundingRect;
+  const anchorCentreX = anchorLeft + anchorWidth / 2;
+  const anchorCentreY = anchorTop + anchorHeight / 2;
+
+  const { width, height } = boundingClientRect;
+
+  let predictedLeft, predictedRight, predictedTop, predictedBottom;
+  if (position === Position.TOP_LEFT) {
```

Comment:
Can't this be changed into a `switch` statement?

Reply:
It can, but is it worth it? There is no `default` case, and every case will have to end with the nasty `break`. What benefit will the switch bring?

Reply:
As discussed offline we'll leave this as it is.